File: solutions/cpp/loops.cpp
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <stdbool.h>
#include <sys/time.h>

#define sampleSize 23

void simulate(int interations);

int main(int argc, char *argv[]) {
  int iterations = atoi(argv[1]);
  simulate(iterations);
  return 0;
}

void simulate(int iterations) {
  struct timeval start, end;
  gettimeofday(&start, NULL);
  srand(start.tv_usec);

  int duplicates = 0;
  for (int x=0; x < iterations; x++) {
    int data[365] = {};
    for (int i=0; i < sampleSize; i++) {
      int number = (float) rand() / RAND_MAX * 365;
      if (data[number] == 1) {
        duplicates++;
        break;
      } else {
        data[number] = 1;
      }
    }
  }
  printf("iterations: %d\n", iterations);
  printf("sample-size: %d\n", sampleSize);
  double results = duplicates * 100.0 / iterations;
  printf("percent: %0.2f\n", results);
  gettimeofday(&end, NULL);
  double startTime = start.tv_sec + start.tv_usec / 1000000.0;
  double endTime = end.tv_sec + end.tv_usec / 1000000.0;
  float diff = endTime - startTime;
  printf("seconds: %.3f\n", diff);
}

File: solutions/python3/loops.py
from random import random
import time
import sys
# import array

def simulate():
    start = time.time()
    iterations = int(sys.argv[1])
    sample_size = 23

    count = 0
    for _ in range(iterations):
        # data = [0] * 365
        data = set() # similar speed
        # data = { -1 } # similar speed
        # data = set([]) # similar speed
        # data = [] # 33% slower
        # data = [None for _ in range(365)] # 3x slower
        # data = array.array('H', [0] * 365) # 3.5x slower

        for _ in range(sample_size):
            rand = int(random() * 365)
            # if data[rand] == 1: # for array and list[] solutions
            if rand in data: # for set and x in list solutions
                count += 1
                break
            else:
                # data[rand] = 1 # for array and list[] solutions
                data.add(rand) # for set solution
                # data.append(rand) # for x in list solution

    print("iterations:", iterations)
    print("sample-size:", sample_size)
    results = round(count / iterations * 100, 2)
    print("percent:", results)
    end = time.time()
    diff = round(end-start, 3)
    print("seconds:", diff)

simulate()
